### PR TITLE
[GH-658] Add golang-ci lint

### PR DESCRIPTION
**Pull request by carantunes**
#### Summary
This PR copies most of the golangci-lint configuration from https://github.com/mattermost/mattermost-plugin-autolink/pull/108 into this repo. This helps with a consistent code style and should improve code quality over time.

Most changes are straightforward and just style wise. I did comment on the more tricky ones.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-plugin-jira/issues/658

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -141,186 +111,63 @@ endif
...
 	@cat Makefile | grep -v '\.PHONY' |  grep -v '\help:' | grep -B1 -E '^[a-zA-Z0-9_.-]+:.*' | sed -e "s/:.*//" | sed -e "s/^## //" |  grep -v '\-\-' | sed '1!G;h;$$!d' | awk 'NR%2{printf "\033[36m%-30s\033[0m",$$0;next;}1' | sort
-
-# sd is an easier-to-type alias for server-debug
-.PHONY: sd
-sd: server-debug
-
-# server-debug builds and deploys a debug version of the plugin for your architecture.
-# Then resets the plugin to pick up the changes.
-.PHONY: server-debug
-server-debug: server-debug-deploy reset
```
Comment (hanzei) [inline Makefile:216]: @jfrerich Just to double check: Are we fine with removing these commands?

File: build/custom.mk [hunk 2]
```diff
@@ -1,5 +1 @@
-# Include custom targets and environment variables here
```
Comment (hanzei) [inline build/custom.mk:1]: Please revert the changes here

File: server/jiratracker/tracker.go [hunk 3]
```diff
@@ -1,4 +1,4 @@
-package jiraTracker
+package jiratracker
```
Comment (hanzei) [inline server/jiratracker/tracker.go:1]: 1/5 Maybe just `tracker` for simplicity?

File: server/user_test.go [hunk 4]
```diff
@@ -54,7 +54,7 @@ func TestRouteUserStart(t *testing.T) {
 	for name, tc := range tests {
 		t.Run(name, func(t *testing.T) {
 			request := httptest.NewRequest("GET", routeUserStart, nil)
-			request.Header.Set("Mattermost-User-Id", tc.userID)
+			request.Header.Set("Mattermost-User-ID", tc.userID)
```
Comment (hanzei) [inline server/user_test.go:57]: That is a predefined header send by the Mattermost Server. It can and should stay as `Mattermost-User-Id`. Otherwise authentication breaks.

**Review by hanzei — CHANGES_REQUESTED** [review]
This is amazing work :tada: Thanks for going through all of these files.

Almost all changes look good. There is one issue with a http header. See my comment below.

I would be fine if you comment out the failing linter and we merge this PR as it is.

File: server/kv.go [hunk 5]
```diff
@@ -369,7 +367,7 @@ func (store store) EnsureRSAKey() (rsaKey *rsa.PrivateKey, returnErr error) {
 
 	if rsaKey == nil {
 		var newRSAKey *rsa.PrivateKey
-		newRSAKey, err = rsa.GenerateKey(rand.Reader, 1024)
+		newRSAKey, err = rsa.GenerateKey(rand.Reader, 2048)
```
Comment (carantunes) [inline server/kv.go:370]: Please doublecheck this change has no effects

File: server/webhook_parser.go [hunk 6]
```diff
@@ -170,8 +169,6 @@ func parseWebhookChangeLog(jwh *JiraWebhook) Webhook {
 
 		if event != nil {
 			events = append(events, event)
-		} else {
-			fieldsNotFound = append(fieldsNotFound, field)
```
Comment (carantunes) [inline server/webhook_parser.go:175]: Please doublecheck this change has no effects, doesn't look like it's used.

**Comment by carantunes** [comment]
@hanzei Ready for review!  🎉

File: server/client.go [hunk 7]
```diff
@@ -370,15 +371,13 @@ var keyOrIDRegex = regexp.MustCompile("(^[[:alpha:]]+-)?[[:digit:]]+$")
 func endpointNameFromRequest(r *http.Request) string {
 	_, path := splitInstancePath(r.URL.Path)
 	l := strings.ToLower(path)
-	s := strings.TrimLeft(l, "/rest/api")
-	if s == l {
-		return "_unrecognized"
-	}
```
Comment (hanzei) [inline server/client.go:376]: Why was this case removed?

File: server/http.go [hunk 8]
```diff
@@ -43,7 +44,7 @@ const (
 	routeACInstalled                            = "/ac/installed"
 	routeACJSON                                 = "/ac/atlassian-connect.json"
 	routeACUninstalled                          = "/ac/uninstalled"
-	routeACUserRedirectWithToken                = "/ac/user_redirect.html"
+	routeACUserRedirectWithToken                = "/ac/user_redirect.html" // #nosec G101
```
Comment (hanzei) [inline server/http.go:47]: :+1:

File: server/issue.go [hunk 9]
```diff
@@ -23,10 +23,21 @@ import (
...
+	labelsField      = "labels"
+	statusField      = "status"
+	reporterField    = "reporter"
+	priorityField    = "priority"
+	descriptionField = "description"
+	resolutionField  = "resolution"
+)
+
+func makePost(userID,
+	channelID, message string) *model.Post {
```
Comment (hanzei) [inline server/issue.go:37]: Nit pick: The formatting look weird. Should this be
```suggestion
func makePost(userID, channelID, message string) *model.Post {
```

File: server/issue.go [hunk 10]
```diff
@@ -1028,10 +1042,12 @@ func (p *Plugin) TransitionIssue(in *InTransitionIssue) (string, error) {
 
 	transitions, err := client.GetTransitions(in.IssueKey)
 	if err != nil {
-		return "", errors.New("We couldn't find the issue key. Please confirm the issue key and try again. You may not have permissions to access this issue.")
+		return "", errors.New(
+
+			"we couldn't find the issue key. Please confirm the issue key and try again. You may not have permissions to access this issue")
```
Comment (hanzei) [inline server/issue.go:1047]: Nit pick:
```suggestion
		return "", errors.New("we couldn't find the issue key. Please confirm the issue key and try again. You may not have permissions to access this issue")
```

File: .golangci.yml [hunk 11]
```diff
@@ -0,0 +1,62 @@
+run:
+  timeout: 5m
+  modules-download-mode: readonly
+
+linters-settings:
+  goconst:
+    min-len: 2
```
Comment (hanzei) [inline .golangci.yml:7]: It might be worth to increase that to 3 and don't `const` strings like `, `. What do you think?

File: server/kv.go (hunk 5 above)
Comment (hanzei) [reply server/kv.go:370]: While I don't think this has an immediate effect, I would like to keep the key length consistent between all instances. Would you please revert the changes?

File: server/user_server.go [hunk 12]
```diff
@@ -59,27 +59,27 @@ func (p *Plugin) httpOAuth1aComplete(w http.ResponseWriter, r *http.Request, ins
...
 		return respondErr(w, http.StatusInternalServerError,
-			errors.WithMessage(appErr, "failed to load user "+mattermostUserId))
+			errors.WithMessage(appErr, "failed to load user "+mattermostUserID))
 	}
 
-	oauthTmpCredentials, err := p.otsStore.OneTimeLoadOauth1aTemporaryCredentials(mattermostUserId)
-	if err != nil || oauthTmpCredentials == nil || len(oauthTmpCredentials.Token) <= 0 {
-		return respondErr(w, http.StatusInternalServerError, errors.WithMessage(err, "failed to get temporary credentials for "+mattermostUserId))
+	oauthTmpCredentials, err := p.otsStore.OneTimeLoadOauth1aTemporaryCredentials(mattermostUserID)
+	if err != nil || oauthTmpCredentials == nil || len(oauthTmpCredentials.Token) == 0 {
```
Comment (hanzei) [inline server/user_server.go:73]: Nit pick: It's more idiomatic to write 
```suggestion
	if err != nil || oauthTmpCredentials == nil || oauthTmpCredentials.Token == "" {
```

File: server/utils/types/id_set.go [hunk 13]
```diff
@@ -42,7 +42,7 @@ func NewIDSet(vv ...ID) *IDSet {
 	return i
 }
 
-func (i *IDSet) Set(v ID) {
+func (i *IDSet) Set(v ID) { //nolint:interfacer
```
Comment (hanzei) [inline server/utils/types/id_set.go:45]: Maybe we just remove `interfacer` for `golangci.yml`? So far it's doesn't been useful.

File: server/webhook_parser.go (hunk 6 above)
Comment (hanzei) [reply server/webhook_parser.go:175]: Look good to me :+1:

File: server/webhook_parser.go [hunk 14]
```diff
@@ -132,54 +130,54 @@ func parseWebhookChangeLog(jwh *JiraWebhook) Webhook {
...
 		}
 	}
-	if len(events) == 0 {
+
+	switch {
```
Comment (hanzei) [inline server/webhook_parser.go:175]: Maybe `switch` on `len(events)`?

**Review by hanzei — CHANGES_REQUESTED** [review]
Awesome! A few comments and then it looks good from my side

File: .golangci.yml (hunk 11 above)
Comment (carantunes) [reply .golangci.yml:7]: I didn't feel like it was my place to suggest it, but I totally agree.

File: server/utils/types/id_set.go (hunk 13 above)
Comment (carantunes) [reply server/utils/types/id_set.go:45]: also agree, seems like the repo was even archived

File: server/client.go (hunk 7 above)
Comment (carantunes) [reply server/client.go:376]: When I changed from TrimLeft to TrimPrefix I broke the tests and it needed a bit of tweaking to work, and after finding the actual problem I never noticed I could re-add it. I've fixed it now, thanks.

File: server/kv.go (hunk 5 above)
Comment (carantunes) [reply server/kv.go:370]: Done :)

File: Makefile (hunk 1 above)
Comment (carantunes) [reply Makefile:216]: @hanzei I did all the requested changes. Only missing clarification on this point. Should I revert this changes? Copied the makefile from autolink as indicated in the issue.

**Review by hanzei — APPROVED** [review]
Amazing work! :+1:

**Comment by jasonblais** [comment]
@jfrerich kind reminder to help with a review of this PR

File: Makefile (hunk 1 above)
Comment (jfrerich) [reply Makefile:216]: I think we should keep and investigate these options.  They were implemented some time ago, but the idea being that we could save time by only building the appropriate architecture.

**Review by jfrerich — APPROVED** [review]
Amazing amount of work :) Thanks, @Adovenmuehle!  

@hanzei, I think we should keep the makefile commands and investigate them. We should consider adding them to the starter-template plugin.

File: Makefile (hunk 1 above)
Comment (hanzei) [reply Makefile:216]: Well, technically the are stored in Git and saved there. 1/5 that consistency is more important that the build features.

**Comment by hanzei** [comment]
@DHaussermann Maybe this is a good PR to test in the next review process

**Comment by DHaussermann** [comment]
Thanks @carantunes for this contribution. I have successfully build and deployed the plugin. I have also done some brief regression testing around core functionality Connect/disconnect create and attach issue etc.. and found no issues.   

 @hanzei regarding additional testing, this would get alot of coverage to catch potential regression when I do the release testing once Jira 3.0.1 is ready and we have a version bump PR. 
 
Do you feel the risk if fairly low here and we can merge this in before the `3.0.1`. release? If so, maybe next steps would be to resolve the conflicts and do one more quick round of testing to get this merged in.

**Comment by hanzei** [comment]
@DHaussermann I highly doubt we will release a 3.0.1 from `master`. There have been a lot of changes on `master` since 3.0.0. _If_ we release a 3.0.0, I would expect it to be a released from a separate branch with cherry-picked commits.

Hence, I think it's fine to merge this PR with low amount of testing and rely on the 3.1.0 release testing to catch any bugs.

@carantunes Could you please merge `master` into your branch?

**Comment by hanzei** [comment]
@DHaussermann Given that 3.0.1 will be not cut from `master` would be fine merging merging this PR as it is?

**Comment by hanzei** [comment]
Awesome work @carantunes :rocket:

**Comment by carantunes** [comment]
Thank you! 🥳  Happy to help again if something comes up!
